pkg/webhook/ippool/mutator.go

```diff
@@ -0,0 +1,180 @@
+package ippool
+
+import (
+	"fmt"
+	"net/netip"
+	"reflect"
+
+	"github.com/harvester/webhook/pkg/server/admission"
+	"github.com/sirupsen/logrus"
+	admissionregv1 "k8s.io/api/admissionregistration/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	networkv1 "github.com/harvester/vm-dhcp-controller/pkg/apis/network.harvesterhci.io/v1alpha1"
+	"github.com/harvester/vm-dhcp-controller/pkg/util"
+	"github.com/harvester/vm-dhcp-controller/pkg/webhook"
+)
+
+const (
+	Start EndpointType = "start"
+	End   EndpointType = "end"
+)
+
+type EndpointType string
+
+type Mutator struct {
+	admission.DefaultMutator
+}
+
+func NewMutator() *Mutator {
+	return &Mutator{}
+}
+
+func (m *Mutator) Create(_ *admission.Request, newObj runtime.Object) (admission.Patch, error) {
+	ipPool := newObj.(*networkv1.IPPool)
+
+	serverIP := ipPool.Spec.IPv4Config.ServerIP
+	if err := ensureServerIP(
+		&serverIP,
+		ipPool.Spec.IPv4Config.CIDR,
+		ipPool.Spec.IPv4Config.Router,
+		ipPool.Spec.IPv4Config.Pool.Exclude,
+	); err != nil {
+		return nil, fmt.Errorf(webhook.CreateErr, "IPPool", ipPool.Namespace, ipPool.Name, err)
+	}
+	pool := ipPool.Spec.IPv4Config.Pool
+	if err := ensurePoolRange(
```

Comment:
we should change the method signature to make the pool object return more specific